src/Microsoft.ML.Data/Utilities/LocalEnvironment.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    using Stopwatch = System.Diagnostics.Stopwatch;
+
+    /// <summary>
+    /// An ML.NET environment for local execution.
+    /// </summary>
+    public sealed class LocalEnvironment : HostEnvironmentBase<LocalEnvironment>
+    {
+        private sealed class Channel : ChannelBase
+        {
+            public readonly Stopwatch Watch;
+            public Channel(LocalEnvironment root, ChannelProviderBase parent, string shortName,
+                Action<IMessageSource, ChannelMessage> dispatch)
+                : base(root, parent, shortName, dispatch)
+            {
+                Watch = Stopwatch.StartNew();
+                Dispatch(this, new ChannelMessage(ChannelMessageKind.Trace, MessageSensitivity.None, "Channel started"));
+            }
+
+            public override void Done()
+            {
+                Watch.Stop();
+                ChannelFinished();
+                base.Done();
+            }
+
+            private void ChannelFinished()
+                => Dispatch(this, new ChannelMessage(ChannelMessageKind.Trace, MessageSensitivity.None, "Channel finished. Elapsed { 0:c }.", Watch.Elapsed));
+
+            protected override void DisposeCore()
```

Comment:
FYI - I logged [https://github.com/dotnet/machinelearning/issues/930](https://github.com/dotnet/machinelearning/issues/930) to implement the Dispose Pattern correctly on these types.